io-sim-classes/src/Control/Monad/Class/MonadProbe.hs

```diff
@@ -60,3 +82,150 @@ instance MonadRunProbe IO IO where
   newProbe  = ProbeIO <$> STM.newTVarIO []
   readProbe (ProbeIO p) = STM.readTVarIO p
   runM = id
+
+
+data LogTrace t' m t = LogTrace {
+      traceOutput :: t -> m ()
+    , readLogTrace :: m t'
+    , traceStopping :: m ()
```

Comment:
What are the second two for? I thought a tracer was just an output mechanism not a query mechanism?

And what is the meaning of the `t'` param?

Reply:
In `SimM`: `traceOutput` will write in a `TVar`and at the end `readLogTrace` will read from it.
Also, `iohk-monitoring-framework`'s `Trace` needs a shutdown.

`t'` is the type of data saved in the `TVar`. To be more specific a msg could be "hello" but we want also to keep the timestamp and the threadId of the sender so `t'` could be `t' :: (ThreadId, Timestamp, String)`.